### PR TITLE
Refactor pgp_key_search_t

### PR DESCRIPTION
**Pull request by ni4**
Another one on the C++-transition way.Thought it would be faster...

### CODE REVIEW

File: src/tests/support.cpp [hunk 1]
```diff
@@ -823,11 +800,7 @@ rnp_tests_key_search(rnp::KeyStore *keyring, const std::string &uid)
     if (!keyring || uid.empty()) {
         return NULL;
     }
-
-    pgp_key_search_t srch_userid(PGP_KEY_SEARCH_USERID);
-    strncpy(srch_userid.by.userid, uid.c_str(), sizeof(srch_userid.by.userid));
-    srch_userid.by.userid[sizeof(srch_userid.by.userid) - 1] = '\0';
-    return keyring->search(srch_userid);
+    return keyring->search(rnp::KeyUIDSearch(uid));
```
Comment (maxirmx) [inline src/tests/support.cpp:803]: Copilot suggests a change to improve coverage :)
```
    auto search = rnp::KeyUseridSearch(uid);
    assert(search.name() == "userid");
    assert(search.value() == uid);
    return keyring->search(search);
```

**Review by maxirmx — CHANGES_REQUESTED** [review]

File: src/lib/key-provider.cpp [hunk 2]
```diff
@@ -24,28 +24,223 @@
...
+    case Type::Grip:
+        if (binval.size() != PGP_KEY_GRIP_SIZE) {
+            RNP_LOG("Invalid grip: %s", value.c_str());
+            return nullptr;
+        }
+        pgp_key_grip_t grip;
+        memcpy(grip.data(), binval.data(), grip.size());
+        return create(grip);
+    default:
+        assert(false);
```
[the remark this answers is not in the document]
Comment (ni4) [reply src/lib/key-provider.cpp:118]: This line to be able to detect possible future coding error. Or do you mean absence of RNP_LOG() for default case?

File: src/tests/support.cpp (hunk 1 above)
Comment (ni4) [reply src/tests/support.cpp:803]: These helper functions are designed to not assert internally - they just to make sure everything is fine on caller level.

File: src/lib/key-provider.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxirmx) [reply src/lib/key-provider.cpp:118]: This line is not covered by tests and assert won't work for release builds.

[the remark this answers is not in the document]
Comment (ni4) [reply src/lib/key-provider.cpp]: @maxirmx Ok, removed.

File: src/tests/support.cpp (hunk 1 above)
Comment (maxirmx) [reply src/tests/support.cpp:803]: Ok, let me say it differently. 

There is a class factory ```std::unique_ptr<KeySearch>KeySearch::create(const pgp_key_id_t &keyid)```  and  there is  limited testing of positive scenarious for it. It is possible to add some extra checks to the helper functions to increase coverage.

**Review by maxirmx — APPROVED** [review]
LGTM

File: src/tests/support.cpp (hunk 1 above)
Comment (ni4) [reply src/tests/support.cpp:803]: @maxirmx Got it, however I'm not sure whether we should test internal classes through internal interfaces, while 'officially' exposed is only FFI interface. I'll update FFI tests with some more edge cases.

**Comment by antonsviridenko** [comment]
> Thought it would be faster...

Is it actually faster now? :P

**Comment by antonsviridenko** [comment]
There is some conflict, probably caused by the recent merged PR...

**Comment by ni4** [comment]
> > Thought it would be faster...
> 
> Is it actually faster now? :P

I mean faster transition to C++ from all the anonymous unions, char * and suchlike internal stuff :) I'll rebase and fix errors and ping you back for review, thanks!

**Review by ronaldtse — APPROVED** [review]
